RFC/src/RFC-0250_Covenants.md

```diff
@@ -0,0 +1,468 @@
+
+This results in an interesting dilemma: how do we allow rules that dictate how future outputs look only armed with
+the knowledge that the rule must apply to one or more outputs?
+
+In this RFC, we propose a covenant scheme that allows the [UTXO] originator to express a _filter_ that must be 
+satisfied for a subsequent spending transaction to be considered valid.
+
+## Assumptions
+
+The following assumptions are made:
+1. Duplicate commitments within a block are disallowed by consensus _prior_ to covenant execution,
+2. all outputs in the output set are valid, and
+3. all inputs are valid spends, save for covenent checks.
+
+## Protocol modifications
+
+Modifications to the existing protocol and consensus are as follows:
+
+- the covenant is recorded in the transaction [UTXO],
+- the covenant is committed to in the output and input hashes to prevent malleability,
+- transactions with covenants entering the mempool MUST be validated, and
+- each covenant in a block must be validated before being included in the block chain.
+
+### Transaction input and output changes
+
+A `covenant` field would need to be added to the `TransactionOutput` and `TransactionInput` structs
+and committed to in their hashes. 
+
+### Covenant definition
+
+We define a clear notation for covenants that mirrors the [miniscript] project.
+
+#### Execution Context and Scope
+
+Covenants execute within a limited read-only context and scope. This is both to reduce complexity (and therefore 
+the possibility of bugs) and maintain reasonable performance. 
+
+A covenant's context is limited to:
+- a immutable reference to the current input,
+- a vector of immutable _references_ to outputs in the current block/transaction (called, the output set), 
+- the current input's mined height, and
+- the current block height.
+
+Each output's covenant is executed with this context, filtering on the output set and returning the result. 
+The output set given to each covenant at execution MUST be the same set for all covenants and MUST never be
+influenced by other covenants. The stateless and immutable nature of this scheme has the benefit of being 
+able to execute covenants in parallel.
+
+A covenant passes if at least one output in the set is matched. Allowing more than one output to match allows for 
+covenants that restrict the characteristics of multiple outputs. A covenant that matches zero outputs _fails_ 
+which invalidates the transaction/block.
+
+If a covenant is empty (zero bytes) the `identity` operation is implied and therefore, no actual execution need occur.
+
+#### Argument types
+
+```rust,ignore
+enum CovenantArg {
+    // byte code: 0x01
+    // data size: 32 bytes
+    Hash([u8; 32]),
+    // byte code: 0x02 
+    // data size: 32 bytes
+    PublicKey(RistrettoPublicKey),
+    // byte code: 0x03
+    // data size: 32 bytes
+    Commitment(PedersonCommitment),
+    // byte code: 0x04
+    // data size: 64 bytes
```

Comment:
Cant we reduce this to 32 bytes by just looking at the hash of a commitment, or output for that matter?
This is stored on the block chain and sent over the wire, we want this to be as small as possible.

Reply:
Yeah that's the general direction I think is good, I think I included this as an "incase we need it" but agree with what you've said